--- exec.c	2018-11-14 22:16:02.916400000 -0600
+++ exec.c.x	2018-07-09 22:29:41.779887000 -0500
@@ -1189,7 +1189,9 @@
 
     qemu_ram_setup_dump(new_block->host, size);
     qemu_madvise(new_block->host, size, QEMU_MADV_HUGEPAGE);
+    /* Keep translated memory blocks across forks for AFL!
     qemu_madvise(new_block->host, size, QEMU_MADV_DONTFORK);
+    */
 
     if (kvm_enabled())
         kvm_setup_guest_memory(new_block->host, size);
