--- translate-all.c	2018-11-14 22:16:03.088381000 -0600
+++ translate-all.c.x	2018-07-09 22:13:09.930511000 -0500
@@ -387,8 +387,13 @@
     /* We can't use g_malloc because it may recurse into a locked mutex. */
 # define ALLOC(P, SIZE)                                 \
     do {                                                \
-        P = mmap(NULL, SIZE, PROT_READ | PROT_WRITE,    \
-                 MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);   \
+      void* _tmp = mmap(NULL, SIZE, PROT_READ | PROT_WRITE, \
+                        MAP_PRIVATE | MAP_ANONYMOUS, -1, 0); \
+      if (_tmp == (void*)-1) { \
+        qemu_log(">>> Out of memory for stack, bailing out. <<<\n"); \
+        exit(1); \
+      } \
+      (P) = _tmp; \
     } while (0)
 #else
 # define ALLOC(P, SIZE) \
